### PR TITLE
Feature flag to disable rmcp sdu length verification

### PR DESCRIPTION
**Pull request by canteuni**
This PR aims to solve the following issue #140 .

It adds a sort of feature flag `check_sdu_length` in `IpmiMsg` constructor, which is enabled by default so the current behavior doesn't change.

When deactivated, this allows to ignore the verification `len(pdu) = header_len + data_len`. This will also return everything contained in the `pdu[header_len:]` array rather than just `pdu[header_len:header_len + data_len]`  .

It only concerns legacy RMCP interface. I agree that the code should not be modified to complain with a buggy IPMI implementation, but I saw that `ipmitool` also ignore this verification when using legacy RMCP, so I thought this could be done here too.

The flag can be used when creating the IPMI interface :
```python
interfaces.create_interface(
    interface="rmcp",
    check_sdu_length=False
)
```

### CODE REVIEW

File: pyipmi/interfaces/rmcp.py [hunk 1]
```diff
@@ -350,7 +360,7 @@ class Rmcp(object):
     _session = None
 
     def __init__(self, slave_address=0x81, host_target_address=0x20,
-                 keep_alive_interval=1):
+                 keep_alive_interval=1, check_sdu_length=True):
```
Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:363]: I think a better solution would be a configuration dict for quirks like this. With that we can avoid having more and more arguments for the interfaces and it is more generic.

e.g. `quirks_cfg={'rmcp_ignore_sdu_length': True}`

Then we can check for that with `if 'rmcp_ignore_sdu_length' in quirks_cfg:`

What do you think?

I agreed please add a doc string for the Rmcp object and describe the usage of that quirk.

File: pyipmi/interfaces/rmcp.py [hunk 2]
```diff
@@ -363,6 +373,7 @@ def __init__(self, slave_address=0x81, host_target_address=0x20,
         self._stop_keep_alive = None
         self._q = Queue()
         self.transaction_lock = threading.Lock()
+        self.check_sdu_length = check_sdu_length
```
Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:376]: here: self.quirks_cfg

Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:376]: `self.ignore_sdu_length = 'rmcp_ignore_sdu_lenght' in quirks_cfg else False`

File: pyipmi/interfaces/rmcp.py [hunk 3]
```diff
@@ -227,8 +227,9 @@ class IpmiMsg(object):
     HEADER_FORMAT_NO_AUTH = '!BIIB'
     HEADER_FORMAT_AUTH = '!BII16BB'
 
-    def __init__(self, session=None):
+    def __init__(self, session=None, check_sdu_length=True):
```
Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:230]: Here invert the log: ignore_sdu_length

File: pyipmi/interfaces/rmcp.py [hunk 4]
```diff
@@ -321,19 +322,28 @@ def unpack(self, pdu):
             (self.auth_type, self.sequence_number, self.session_id,
                 data_len) = struct.unpack(self.HEADER_FORMAT_NO_AUTH, header)
 
-        if len(pdu) < header_len + data_len:
-            raise DecodingError('short SDU')
-        elif len(pdu) > header_len + data_len:
-            raise DecodingError('SDU has extra bytes ({:d},{:d},{:d} )'.format(
-                len(pdu), header_len, data_len))
+        if self.check_sdu_length:
```
Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:325]: `if not self.ignore_sdu_lenght`

File: pyipmi/interfaces/rmcp.py (hunk 3 above)
Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:230]: Here invert the log: `ignore_sdu_length=False`

File: pyipmi/interfaces/rmcp.py [hunk 5]
```diff
@@ -227,8 +227,9 @@ class IpmiMsg(object):
     HEADER_FORMAT_NO_AUTH = '!BIIB'
     HEADER_FORMAT_AUTH = '!BII16BB'
 
-    def __init__(self, session=None):
+    def __init__(self, session=None, check_sdu_length=True):
         self.session = session
+        self.check_sdu_length = check_sdu_length
```
Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:232]: `self.ignore_sdu_length = ignore_sdu_length`

Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:232]: `self.ignore_sdu_length = ignore_sdu_length`

File: pyipmi/interfaces/rmcp.py [hunk 6]
```diff
@@ -387,11 +398,11 @@ def _send_ipmi_msg(self, data):
         tx_data = ipmi.pack(data)
         self._send_rmcp_msg(tx_data, RMCP_CLASS_IPMI)
 
-    def _receive_ipmi_msg(self):
+    def _receive_ipmi_msg(self, check_sdu_length=True):
```
Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:401]: ignore_sdu_length=False

File: pyipmi/interfaces/rmcp.py [hunk 7]
```diff
@@ -387,11 +398,11 @@ def _send_ipmi_msg(self, data):
...
         (_, class_of_msg, pdu) = self._receive_rmcp_msg()
         if class_of_msg != RMCP_CLASS_IPMI:
             raise DecodingError('invalid class field in ASF message')
-        msg = IpmiMsg()
+        msg = IpmiMsg(check_sdu_length=check_sdu_length)
```
Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:405]: msg = IpmiMsg(ignore_sdu_length=ignore_sdu_length)

Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:405]: msg = IpmiMsg(ignore_sdu_length=ignore_sdu_length)

File: pyipmi/interfaces/rmcp.py [hunk 8]
```diff
@@ -562,7 +573,7 @@ def _send_and_receive(self, target, lun, netfn, cmdid, payload):
                 if not self._q.empty():
                     rx_data = self._q.get()
                 else:
-                    rx_data = self._receive_ipmi_msg()
+                    rx_data = self._receive_ipmi_msg(self.check_sdu_length)
```
Comment (hthiery) [inline pyipmi/interfaces/rmcp.py:576]: self.ignore_sdu_length

**Review by hthiery — COMMENTED** [review]
This is a proposal how it could be implemented in a more general way.

File: pyipmi/interfaces/rmcp.py (hunk 1 above)
Comment (canteuni) [reply pyipmi/interfaces/rmcp.py:363]: Yeah you're right, it is more "future proof" this way :+1: 
I will add doc string and make the corresponding modifications

**Comment by hthiery** [comment]
Looks good. Thank you!

**Pull request merged**